ARROW-1073: C++: Adapative integer builder

xhochy opened the pull request.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+            [](int64_t x) { return static_cast<int32_t>(x); });
+      } break;
+      default:
+        DCHECK(false);
```
What should happen in release builds?

xhochy replied:
This code path should never be reached. Without the `default:` section, we would have a compiler warning about unhandled values.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      reinterpret_cast<int64_t*>(raw_data_)[length_++] = val;
+      break;
+    default:
+      DCHECK(false);
```
Release builds?

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      RETURN_NOT_OK((ExpandIntSizeN<int64_t>()));
+      break;
+    default:
+      DCHECK(false);
```
dchecks

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      break;
+    default:
+      DCHECK(false);
+      return Status::NotImplemented("Only ints of size 1,2,4,8 are supported");
```
DCHECK or only error status?

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+}
+
+// Check if we would need to expand the underlying storage type
+uint8_t expanded_uint_size(uint64_t val, uint8_t current_int_size) {
```
static inline?

xhochy replied:
As this is a private method, this should not be needed, the compiler should automatically take care of inlining.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      RETURN_NOT_OK((ExpandIntSizeN<uint32_t>()));
+      break;
+    case 8:
+      RETURN_NOT_OK((ExpandIntSizeN<uint64_t>()));
```
It might be overkill, but you could potentially collapse these classes with an extra template parameter that switches between signed and unsigned variants. So something like

`class AdaptiveIntBuilder : AdaptiveIntBuilderBase<AdaptiveIntBuilder>`

and then you could define various static attributes on `AdaptiveIntBuilder`

xhochy replied:
I though about that and did not see yet how this would reduce code complexity while keeping it understandable.

xhochy commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+}
+
+// Check if we would need to expand the underlying storage type
+uint8_t expanded_int_size(int64_t val, uint8_t current_int_size) {
```
See below; `inline` is mostly these days only about linkage and visibility and not so much about if this function should really be inlined.

xhochy commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+      case 1: {
+        int8_t* data_ptr = reinterpret_cast<int8_t*>(raw_data_) + length_;
+        std::transform(values, values + length, data_ptr,
+            [](int64_t x) { return static_cast<int8_t>(x); });
```
In the background it should actually be that. The only differentiation to a plain for-loop except the savings in lines of code is that with C++17 you could have the compiler/STL run the loop multi-threaded.

xhochy commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+
+#define __LESS(a, b) (a) < (b)
+template <typename new_type, typename old_type>
+typename std::enable_if<__LESS(sizeof(old_type), sizeof(new_type)), Status>::type
```
There is `std::less` but that also didn't work here.

xhochy commented on cpp/src/arrow/builder.cc:
```diff
@@ -248,6 +249,429 @@ template class PrimitiveBuilder<HalfFloatType>;
+  if (current_int_size == 8 ||
+      (current_int_size < 8 &&
+          (val > static_cast<int64_t>(std::numeric_limits<int32_t>::max()) ||
+              val < static_cast<int64_t>(std::numeric_limits<int32_t>::min())))) {
```
It could be but I'm neither sure which pattern to use and benchmarks are showing that this is not really relevant performance-wise.

xhochy commented:
Fixed an unchecked status and added a benchmark for unsigned integers. Performance-wise, the builders are clearly dominated by the expand & copy logic, thus I opened a JIRA to discuss if we should make `jemalloc` the default. I would leave the DCHECKs as-is because they only are check internal things and the code paths are actually unreachable but there so that the compiler does not complain.

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -247,6 +247,104 @@ using HalfFloatBuilder = NumericBuilder<HalfFloatType>;
+  using ArrayBuilder::Advance;
+
+  /// Scalar append
+  Status Append(uint64_t val);
```
Does inlining scalar appends make much of a perf impact?

wesm commented:
I think this is OK except for me comment about inlining. I suppose I can play with the benchmarks after this goes in

xhochy commented:
Benchmarks about the scalar append:

```
BM_BuildAdaptiveIntNoNulls/repeats:3                            58 ms         57 ms          9      1114MB/s
BM_BuildAdaptiveIntNoNulls/repeats:3                            57 ms         57 ms          9    1.1025GB/s
BM_BuildAdaptiveIntNoNulls/repeats:3                            57 ms         57 ms          9   1.10328GB/s
BM_BuildAdaptiveIntNoNulls/repeats:3_mean                       57 ms         57 ms          9   1124.24MB/s
BM_BuildAdaptiveIntNoNulls/repeats:3_stddev                      0 ms          0 ms          0   7.24806MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3               115 ms        114 ms          6   560.072MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3               114 ms        114 ms          6   562.683MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3               114 ms        114 ms          6   563.081MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean          114 ms        114 ms          6   561.945MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_stddev          0 ms          0 ms          0   1.33449MB/s
-- Append inlined --
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3                80 ms         80 ms          9   799.207MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3                80 ms         80 ms          9   800.331MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3                83 ms         82 ms          9   782.643MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_mean           81 ms         81 ms          9   794.061MB/s
BM_BuildAdaptiveIntNoNullsScalarAppend/repeats:3_stddev          1 ms          1 ms          0   8.08636MB/s
```

So it seems inlining it makes a significant difference. I will make a clean implementation of that.

xhochy commented:
Green Appveyor build: https://ci.appveyor.com/project/xhochy/arrow/build/1.0.372

wesm commented on cpp/src/arrow/builder-benchmark.cc:
```diff
@@ -61,4 +61,70 @@ static void BM_BuildVectorNoNulls(
+  while (state.KeepRunning()) {
+    AdaptiveIntBuilder builder(default_memory_pool());
+    for (int64_t i = 0; i < size; i++) {
+      builder.Append(data[i]);
```
To make this benchmark more "realistic" we should add an assertion about the result of `Append` being OK
